### ISSUE TITLE
Suggestion: Bounty Program

### ISSUE THREAD

**Issue by nunodonato** [godotengine/godot#3792]
(sorry if GitHub is not the best place for this, but I dont know where else to post it)

Would be great if we could have a community Bounty Program, as many other free software programs do or have done in the past.

Basically it would work like this:
1 - Godot's team (or any individual person) can select a bug or new feature that they would like to have fixed/implemented (condition: it must exist in the github issues and must be approved by godot's team has something to be fixed/implemented)
2 - Any person can chip in to that bug/feature
3 - once any developer is happy with the amount raised, and he successfully completed it (PR accepted) he would receive the sum

(one other option would be to have the developer himself announce how much he would need to fix it)

Discuss!  :sunglasses:

**Comment by bojidar-bg** [comment]
GitHub integrated service that might be used - https://www.bountysource.com/

**Comment by Ace-Dragon** [comment]
The tricky part with bounty programs is that there's a real chance the developer could leave the scene completely once his work is done,meaning that any issues caused by the feature has to be fixed by other developers.

More likely, you would also need to require the developer to stick around to fix any issues that will inevitably be the result of his feature or fix.

**Comment by TeoTwawki** [comment]
@Ace-Dragon if you look at it that way pull requests may as well not exist

**Comment by Ace-Dragon** [comment]
The difference between pull requests and bounties is that people who make the requests have a genuine interest in the project (to the point where they don't mind contributing as a volunteer if needed).

Perhaps the bounty system could be geared towards paying people who have already contributed a bit to Godot's codebase, it would certainly reduce the length of time between the initial offer and a tangible contribution (due to less learning time required beforehand).

**Comment by TeoTwawki** [comment]
If a project merges code that absolutely depends on its original author, it has problems way bigger than my interest level in that project.

**Comment by Ace-Dragon** [comment]
Who would you trust more to make a paid submission that is highly beneficial to a large chunk of Godot's users and also respects the innovative design of the Godot software?

- The people who currently have pending requests?
https://github.com/godotengine/godot/pulls

- Or a person with some currently unknown background who discovered an opportunity to make some money?

**Comment by TeoTwawki** [comment]
You really aren't getting it: *If its not up to standard,* ***don't accept it.***

You seem to be telling me that you would accept maintainable code just because the person contributing it *might* be around to try and fix it down the road. You also seem to think negatively of anyone working to fix a bug off of a "bounty". A bug is a bug, a code is code, and if the code is good and maintainable all else is irrelevant. Or perhaps you think someone accepting the bounty reward for fixing something has done something wrong or somehow lessened the contributions of others who did not, making this whole argument purely emotional.

I don't even care if a bounty system were used or not, I simply felt the need to point out the same standards should apply if one is used and code needs to be maintainable no matter the source. I have contributed to many projects over the years, I am usually not around years later. I'm never gotten paid for my open source work, which only a small fraction of is on github. This would be a *terrible* criteria to judge my potential contributions on. Please tell me you do not intend to actually debate this through posts to a github issue suggestion. 

If you want to reward contributions in general, a bounty type system is not ideal. This is 100% unrelated to whether or not the code will be maintained in the future. Expecting the original contributor to be there later is not a good idea - things happen. I could submit code that gets merged today then get hit by a bus tomorrow, then what?

This project caught my interest this week, so I have started browsing its issues to see if there is anything I might be interested in helping with. This doesn't mean I'll still be here next year. If that is a criteria for contributing, then sorry but I'm out.

And yes if a bounty system did get implemented, I would be more likely to choose to fix an issue with an attached bounty instead of "whatever I feel like today". I usually code what I feel like coding if I believe I am competent enough for the task, or have at least an opportunity to learn and improve my skills while doing so.

**Comment by Ace-Dragon** [comment]
Well Godot is not my project for starters, the opinions that would carry the most weight would come from either Reduz or Punto. It's just my idea that suggestions related to development should be looked at from all angles (which includes anything that might not make itself known until sometime in the future).

**Comment by fluffrabbit** [comment]
A bounty system may be a conflict of interest. Imagine a vendor funding support of their product over others, even prioritizing that over things more central to Godot itself. Businesses have more money than a handful of individual donors. I've seen a couple of open source projects reject the idea of a bounty system completely, as well as all forms of advertisement, for fear of being controlled. Conflict of interest, I say.

**Comment by vnen** [comment]
Well, you can already donate to Godot to support development: https://godotengine.org/donate

I know this is not what was suggested, but I prefer that funds go to the Godot organization through Conservancy and the core developers decide which are the top priority features to work on.

IMO what people most want is seldom what is actually best for the engine.

**Comment by punto-** [comment]
I don't think it's there's a conflict of interest problem. If a vendor
wants to fix all the bugs for their platform, it just means there'll be one
platform better supported than the others, but it doesn't make the others
worse. If the others want to have good support for their platform, they can
put up their own money to fix it. If there's a situation where a patch
benefits a vendor and damages the quality of the rest of the engine, we can
intervene.

**Comment by fluffrabbit** [comment]
@punto- The Godot developers' needs need to be met sure enough, and if I were one of them I would love the opportunity to be paid to write code for some company without actually being an employee thereof, but as a Godot user I can't help but be bothered by this possibility, being that Godot is supposed to be community-centric ultra-permissive FLOSS without any commerial allegiance except to Okam Studio. It's human nature to put money before integrity, and maybe the value of integrity isn't all that great, but this doesn't sit well with me. I'm sure NVidia or Valve or whoever will probably branch Godot and put their own developers on it, and maybe some current Godot devs will join the other team, but I would hope that Godot keeps its integrity for at least a couple more years while it has a chance to expand equally on what it already has, fix bugs, and continue down the roadmap.

**Comment by indolering** [comment]
I'm really surprised by the negativity in this thread!    The developers are still in control of what gets merged into the codebase, so quality isn't an issue.

In my experience, BountySource doesn't attract one-off contributions from random contractors.  It's primarily a way to coalesce funding from multiple sources and the bounties tend to go towards core developers.

Fears of financial incentives distorting development are naive: Godot exists because the original developers were making a living from selling commercial games.  Greater use of Godot in commercial products means more funding for development.

BountySource is a great way to convert bugs and feature requests into donations.  I strongly suggest the project adopt it.

**Comment by fluffrabbit** [comment]
>The developers are still in control of what gets merged into the codebase, so quality isn't an issue.

Core Godot devs would no doubt partake in the bounty system. After all, it's their project. Code quality would be consistent with what's already there (not saying much), and the features and fixes suggested by donors would get met, what few of them are so difficult that financial incentive would be necessary.

>BountySource

You just name-dropped a website. I'm sure the fruity European neckbeards who started Godot have already taken a look and are forming their opinions privately.

>convert bugs and feature requests into donations

Therein lies my main concern. What is a bug and what is a feature? There is currently a backlog of 1,361 issues and 90 unmerged pull requests. The userbase (which I won't qualify as a "community") is much larger than the number of core developers, which is understandable considering it's a relatively young formerly in-house program. Democratizing development is a key step to getting as big as Blender, but how do we do that? Github is a good start if an unskilled and abrasive programmer like me can bitch about issues and see the consensus of other users. The bureaucracy works. Will the next bureaucracy also work?

In my opinion, Godot 2.1 is robust and production-ready. It delivers a phat set of stable features which are needed by everyone who uses it, and it just barely meets their needs, but sure enough it does meet them and the core devs understand that this is a product that needs to work, which is head and shoulders above the design philosophy of most other open source projects.

BUT, there are still bugs. Most of these bugs limit the experience on mobile platforms, and some of them are just plain mysterious. Paying cash to get a bug fixed is like paying protection to the mob. If a user (hopefully not me) were chugging away at a project for 6 months and suddenly ran into a show-stopping bug, would that user have to sell the right leg of his grandmother's corpse to get it fixed?

Features are like bugs sometimes. For instance, if a user is developing a VR game for iOS and finds out that the only sensor currently supported on that platform is the accelerometer, that user is likely to get more than a little upset. Thankfully, anyone who can afford both a Mac and an iPhone can probably pay a bounty for it, but I can't.

The question is not where the bounties are going but where they are coming from. But like I was saying about Godot 2.1, I think overall Godot is currently good enough for making a very simplistic 3D game, which is all some people want, so further developments can't be all that bad. If I were putting as much time into game projects as I used to I would be outraged over this shit. Hacktoberfest is something I cannot partake in, but I'm sure it will do some good. Stagnation much? Ehh, the bounty system is a mixed bag. On the one hand it shifts focus to the detriment of Godot's all-around usefulness, on the other it increases development because someone with the cash can pay to get things done.

Sorry, my brain don't work so well these days. I hope my comments have contributed to the discussion.

**Comment by akien-mga** [comment]
Just for the reference:

> You just name-dropped a website. I'm sure the fruity European neckbeards who started Godot have already taken a look and are forming their opinions privately.

Godot was created by Argentinians...

> 90 unmerged pull requests

That's just 35 more than last week because the community is crazy active, and it takes time to review changes before merging them. It will be back to ~60 pretty soon. And in those 60 there are tons of things that can't be merged but haven't been closed yet because there's still some valuable code in there... but they'll end up closed, merged or reimplemented for 3.0 if they can.

Just giving some input so that it's not seen as "Godot devs don't care, they let PRs pile up and never merge them". We've merged more than 2000 PRs over two years, that's incredible for an open source project.

**Comment by akien-mga** [comment]
About having a bounty program, I don't think it means you would *have to* pay to get this very annoying bug that blocks your project fixed. It means you *will be able* to propose money to speed it up. But regarding the examples you give, money or not, they won't implement themselves automagically.

What's important is the mindset of the community. If some developers start to work only on what is monetized, then yeah, it would suck, because some important issues might be less visible and have less money on them. Right now devs work on what they can, want, and feel is the most important for the community. I don't think that would change if bounties arise, but that's still a valid concern.

**Comment by fluffrabbit** [comment]
>they won't implement themselves automagically

Isn't that the point of an issue tracker, or an incentive like Hacktoberfest? Money is an incentive too, and if that's as deep as we look at it then the only sane option would be to implement a bounty system.

If every single issue had a dollar amount on it somehow based on the exact amount of developer time it would take the average hobbyist programmer, what issues get tackled wouldn't be affected much and there would be more developers on the project.

The reality of course is that some issues would have bounties and some wouldn't.

>If some developers start to work only on what is monetized, then yeah, it would suck, because some important issues might be less visible and have less money on them.

I think that describes a lot of people, at least people I hang around. Hitmen, prostitutes, etc. Money does a lot.

**Comment by fluffrabbit** [comment]
I'm probably going to get banned for saying this, but I couldn't resist.

In Star Wars, Leia says to Han "If money is all you care about, then that is all you shall receive." Han should have said, "What, as opposed to getting my dick sucked?" And that's when Luke gave him that sly grin.

This issue is tagged with "discussion". I kind of wish it weren't. The suggestion at first glance, at least to me, seems corrupt and underhanded, and, at least to me, the obvious thing is the prioritization of issues to the highest bidder, while anybody who doesn't donate is told to fuck off. But there may be merit to the suggestion as well, and different bounty websites should be considered, not just the first one that is suggested. (Who knows where these folks work?) In the end, the decision seems to be up to Juan and Ariel with input from Akien and one other person whose name I'm forgetting. I like getting shit done, but the way to get shit done with regards to Godot is questionable. So it's all up to the core devs whose baby this is. Not my baby.

**Comment by TeoTwawki** [comment]
Oh for crying out loud.

**Comment by TeoTwawki** [comment]
@akien-mga 

> What's important is the mindset of the community. 

If this is any indication, your community has abandoment issues or paranoia. Suggest closing issue unless this is an actual seriosu consideration. No sense letting their imaginations run wild if it isn't. Meanwhile, I'll be sure to unsubscribe from the issue and not ever be apart of said community. Good luck with your project.

**Comment by neikeq** [comment]
@fluffrabbit the issue is labelled with _"discussion"_ like any other issue that needs to debate whether something should be made or not.

**Comment by indolering** [comment]
> Core Godot devs would no doubt partake in the bounty system. After all, it's their project. Code quality would be consistent with what's already there (not saying much)

...

> I'm sure the fruity European neckbeards who started Godot have already taken a look and are forming their opinions privately.

I would suggest y'all get a community code of conduct, this kind of stuff is toxic.

**Comment by akien-mga** [comment]
So, so far this issue brought people about who have never participated in this community and are prompt to judge it (money, heh), and our usual trolling and disrespectful bunny.

Closing as this is completely pointless. When we want to discuss bounty programs seriously, we'll do that on IRC with people who do care about Godot's development, not with random people on a GH issue.

The code of conduct of the Godot community is implict: Don't be a jerk. For some, it's hard to follow, but you can't really ban people from GitHub, can you?
